docs/usage/config.md
# Configuration

There are two primary ways in which nendo can be configured:

**Passing a `NendoConfig` upon nendo initialization**

Inside your python app:

```python
from nendo import Nendo, NendoConfig

nd = Nendo(
    config=NendoConfig(
        library_path="./my_library",
        log_level="debug",
        # more configuration variables ...
    )
)
```

**Using environment variables**

Inside the shell, before starting Python, execute for each desired configuration variable:

```bash
$ export LIBRARY_PATH=./my_library
$ export LOG_LEVEL=debug
# more configuration variables ...
```

## Configuration Reference

| **NendoConfig** | **Env var** | **Type** | **Default** | **Description** |
|---|---|---|---|---|
| log_level | LOG_LEVEL | `str` | `"info"` | The log level with which the nendo logger runs. |
| log_file_path | LOG_FILE_PATH | `str` | `""` | The path to where the nendo log should be saved. If none is given (empty string), print to `stdout` |
| plugins | PLUGINS | `List[str]` | `[]` | List of plugins package names to be loaded with Nendo. |
| library_plugin | LIBRARY_PLUGIN | `str` | `"default"` | The name of the nendo library plugin to use. Typically, its name follows the pattern `nendo_plugin_library_[name]`, where `[name]` is the name of the database backend. If set to `"default"`, the default [DuckDB](https://duckdb.org/) implementation of the [NendoLibrary](library.md) will be used. |
| library_path | LIBRARY_PATH | `str` | `"nendo_library"` | The path to the directory to be used for storing the nendo Library files. |
| user_name | USER_NAME | `str` | `"nendo"` | The name of the nendo user to be used for the [NendoLibrary](library.md). Only relevant if deploying nendo together with an API server. |
| user_id | USER_ID | `str` | `"ffffffff-1111-2222-3333-1234567890ab"` | The user ID of the default user to be used for the [NendoLibrary](library.md). Only relevant if deploying nendo together with an API server. |
| auto_resample | AUTO_RESAMPLE | `bool` | `False` | Flag that determines whether tracks should be automatically resampled upon import. |
| default_sr | DEFAULT_SR | `int` | `44100` | The default sample rate to be used when auto-resampling tracks upon import. |
| copy_to_library | COPY_TO_LIBRARY | `bool` | `True` | Flag that determines whether an imported track's file should be copied into the nendo library. |
| auto_convert | AUTO_CONVERT | `bool` | `True` | Flag that determines whether an imported track's file should be converted to Nendo's standard file format (`.wav`). |
| skip_duplicate | SKIP_DUPLICATE | `bool` | `True` | Flag that determines whether a track that points to a file that already exists in the library can be important multiple times. If True, always the file that already exists in the library will be used instead. |
| replace_plugin_data | REPLACE_PLUGIN_DATA | `bool` | `False` | Flag that determines whether plugin data should be replaced, if for a specific plugin name, version and key, two values are added consecutively via `track.add_plugin_data()` or `nendo.library.add_plugin_data()`. If `True`, the second call will cause Nendo to overwrite the existing value for the given plugin name, version and key combination. If `False`, the plugin data will be added in addition to the existing one. Defaults to `False`. |
| max_threads | MAX_THREADS | `int` | `2` | Maximum number of threads to be used for multiprocessing tasks. |
| batch_size | BATCH_SIZE | `int` | `10` | Batch size to use for multiprocessing tasks. |
| stream_mode | STREAM_MODE | `bool` | `False` | Flag that enables `stream mode`: With stream mode, all functions that return multiple items, such as e.g. `nd.get_tracks()` return an `Iterator` instead of a `List`. |
| stream_chunk_size | STREAM_CHUNK_SIZE | `int` | `1` | Size of the chunks (in items) in which the `Interator`s will give back the results. Ignored if `stream_mode` is `False`.

docs/platform/config.md
# Configuration

!!! tip "Running on localhost"
    When running Nendo Platform only on your local machine, you don't need to change any configuration parameters. Nendo Platform will run just fine with the default settings.

You can configure Nendo Platform by setting various configuration parameters. From the shell in which you are starting Nendo Platform, set any of the following environment variables that will be picked up when you call any of the `make` commands.

!!! example "Example: Enable SSL"
    ```bash
    export USE_SSL=true
    ```


**Environment Variable** | **Type** | **Default** | **Description** |
|---|---|---|---|
USE_GPU | `bool` | `true` | Flag that switches between GPU enabled and CPU-only mode. |
SERVER_URL | `str` | `"http://localhost"` | URL under which the server will be made available online. |
USER_STORAGE_SIZE | `int` | `-1` | Storage limit for the user's library in kb. If set to -1, no storage limit will be applied. |
USE_SSL | `bool` | `false` | Flag to switch SSL on or off. Make sure to also set `SSL_CERTIFICATE_PATH` and `SSL_KEY_PATH` or copy your certificate/key combination to the `./conf/nginx/certs/` folder. |
SSL_CERTIFICATE_PATH | `str` | `"./conf/nginx/certs/nendo.crt"` | Path to the SSL certificate file (only relevant when `USE_SSL=true`). |
SSL_KEY_PATH | `str` | `"./conf/nginx/certs/nendo.key"` | Path to the SSL key file (only relevant when `USE_SSL=true`). |
MAILGUN_FROM_ADDRESS | `str` | `"postmaster@yourdomain.com"` | Sets the `FROM` for administrative emails (registration, verification, password reset, etc.) from the server. |
MAILGUN_API_KEY | `str` | `"REPLACE_KEY"` | Mail API key. |


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!-- insertion marker -->
## [0.2.0](https://github.com/okio-ai/nendo/releases/tag/0.2.0) - 2024-02-19

- Added the first extension for the Nendo Library: The `NendoLibraryVectorExtension` is a mix-in class that can be used by implementations of the Nendo Library to add support for saving and retrieving embedding vectors.
- Added a new class `NendoEmbedding` that represents an embedding of a `NendoTrack` and stores the vector together with the ID of the track from which it was computed, the text representation of the track that was used to compute the embedding, as well as the embedding plugin's name and version that were used to compute the embedding.
- Added a new subclass of the `NendoPlugin` called `EmbeddingPlugin` that accepts either a `NendoTrack`, a `NendoCollection`, a `signal` and a `text`, or only a `text` and computes corresponding `NendoEmbedding`(s), depending on whether a single object or a `NendoCollection` was provided. It also determines whether the currently used `NendoLibrary` implementation uses the `NendoLibraryVectorExtension` and, if so, saves the computed embeddings directly to the library. 
- Fixed a problem with the way dictionaries (like `NendoTrack.meta` and `NendoTrack.resource`) were stored in the Nendo Library which resulted in them being stored as JSON strings instead of JSON dictionaries. **This means that upgrading Nendo will break any apps that use libraries that were created with previous versions of nendo.** Please make sure you understand the implications and either flush your DB or convert it to match the new data model.
- Added the global configuration variable `replace_plugin_data`, that specifies whether new plugin data will overwrite existing plugin data for the specific plugin name and version used to generate the data.
- Reworked the way relationships between tracks are managed. Previously, Nendo would create bidirectional relationships, which caused problems with the retrieval of related tracks, especially when using paging (i.e. `offset` and `limit`). The new approach is to only store relationships _from_ a derivative track _to_ the original track from which it was derived. Accordingly, all functions for retrieving related tracks accept a new parameter, `direction: str`, which can assume either one of the values `"to"`, `"from"` and `"both"` and will change their retrieval behavior accordingly. Also, the method names for some methods were changed to reflect the new semantics more accurrately: `NendoTrack.has_relationship_to()` was replaced by `NendoTrack.has_related_track(direction=...)`. `collection.has_relationship_to()` was replaced by `collection.has_related_collection()`.
- Added a new shortcut function `NendoTrack.relate_to_track()` that creates a relationship of the specified `relationship_type` and with the specified metadata from the given track _to_ another track.
- Fixed a bug where the length of a `NendoTrack` would not be correctly determined if the track had a mono signal.
- Two new shortcut functions `NendoTrack.refresh()` and `collection.refresh()` were added that retrieve the latest version of the given object as it exists in the Nendo Library, effectively allowing to quickly pull the latest changes to an object from the database.
- The `NendoTrack.get_plugin_data()` methods signature was extended to allow for filtering a track's plugin data also by `plugin_version` and `user_id` in addition to the existing filters.
- Adjusted the `add_track()` function:
    - Stores all the ID3 metadata in `NendoTrack.meta` instead of `NendoTrack.resource.meta`.
    - If no ID3 metadata was found `NendoTrack.meta.title` will be set to the original name of the imported file.
- Added the `load_related_tracks: bool = False` flag to the `NendoLibrary.get_tracks()` function that allows to enable/disable the populating of the `NendoTrack.related_tracks` field upon retrieval of many tracks from the Nendo Library. This optimizes the performance of the call and allows to retrieve smaller objects that are more suitable for network transmission in web applications that use Nendo Core.
-  Add support for `NendoLibrary.filter_tracks(collection_id="...", order_by="collection")` that allows to order of the retrieved tracks by their `relationship_position` in the referenced collection.
- Added the `NendoLibrary.filter_related_tracks()` method to allow for filtering of related tracks queries.
- Changed the `NendoLibrary.filter_tracks()` method to search over `NendoTrack.meta` in addition to `NendoTrack.resource`. Changed the name of the `resource_filters` argument to `search_meta` to reflect the new behavior.
- Added the `NendoLibrary.add_tracks_to_collection()` function to allow for multiple tracks given as a list of IDs to be added to the specified collection.
- Added the `order` parameter to the `NendoLibrary.get_collection_tracks()` function to allow for control of the ordering. Defaults to `"asc"`, which will order the collection's tracks in ascending order of their `relationship_position`.
- Changed the signature and behavior of the `get_collection()` function:
    - Renamed the `details` parameter to `get_related_tracks` to better reflect it's effect on the function. The parameter controls whether the `NendoCollection.related_tracks` field will be populated.
    - The function now always returns a `NendoCollection` instead of returning a `NendoCollectionSlim` if `details=False` as is was previously.
- Added the `user_id` parameter to the following functions:
    - `NendoTrack.local()`
    - `NendoTrack.get_plugin_value()`
    - `NendoLibrary.get_track()`
    - `NendoLibrary.remove_collection()`

src/nendo/config.py
"""Settings used to configure nendo."""
from enum import Enum
from functools import lru_cache
from typing import List

from pydantic import Field
from pydantic_settings import BaseSettings


class Environment(Enum):
    """Enum used to denote the environment in which nendo is running."""

    LOCAL = "local"
    TEST = "test"
    REMOTE = "remote"


class NendoConfig(BaseSettings):
    """The basic nendo configuration object.

    Uses pydantic's settings management features. Refer to the
    [documentation page on configuration management](../usage/config.md) for
    a full list of configuration options.
    """

    environment: Environment = Field(default=Environment.LOCAL)
    log_level: str = Field(default="WARNING")
    log_file_path: str = Field(default="")
    plugins: List[str] = Field(default_factory=list)
    library_plugin: str = Field(default="default")
    library_path: str = Field(default="nendo_library")
    user_name: str = Field(default="nendo")
    user_id: str = Field(default="ffffffff-1111-2222-3333-1234567890ab")
    auto_resample: bool = Field(default=False)
    default_sr: int = Field(default=44100)
    copy_to_library: bool = Field(default=True)
    auto_convert: bool = Field(default=True)
    skip_duplicate: bool = Field(default=True)
    replace_plugin_data: bool = Field(default=False)
    max_threads: int = Field(default=2)
    batch_size: int = Field(default=10)
    stream_mode: bool = Field(default=False)
    stream_chunk_size: int = Field(default=1)


@lru_cache()
def get_settings() -> NendoConfig:
    """Return the Nendo configuration, cached."""
    return NendoConfig()
